packages/deployments/contracts/deploy/deploy.ts

```diff
@@ -1,48 +1,90 @@
-import { DeployFunction } from "hardhat-deploy/types";
+import { DeployFunction, DeploymentSubmission } from "hardhat-deploy/types";
 import { Contract, Signer } from "ethers";
 
 import { NOMAD_DEPLOYMENTS, SKIP_SETUP, WRAPPED_ETH_MAP } from "../src/constants";
 
+const saveDeployment = async (
+  address: string,
+  name: string,
+  args: any[],
+  artifactName: string,
+  hre: HardhatRuntimeEnvironment,
+) => {
+  const artifact = await hre.deployments.getExtendedArtifact(artifactName);
+  const deploymentSubmission: DeploymentSubmission = {
+    address: address,
+    args: args,
+    ...artifact,
+  };
+
+  hre.deployments.save(`${name}`, deploymentSubmission);
+};
 // Helper function
-const deployUpgradeable = async <T extends Contract = Contract>(
+const deployBeaconProxy = async <T extends Contract = Contract>(
   name: string,
   args: any[],
-  controllerAddress: string,
   deployer: Signer & { address: string },
   hre: HardhatRuntimeEnvironment,
 ): Promise<T> => {
-  // Get init data
   const factory = await hre.ethers.getContractFactory(name, deployer.address);
+  const proxyName = `${name}UpgradeBeaconProxy`;
   const initData = factory.interface.encodeFunctionData("initialize", args);
 
-  // Deploy implementation
-  const implementationDeployment = await hre.deployments.deploy(name, {
-    from: deployer.address,
-    log: true,
-  });
-  const implementationAddress = implementationDeployment.address;
+  // Check if already deployed. then upgrade
+  const proxyDeployment = await hre.deployments.getOrNull(proxyName);
 
-  // Deploy beacon
-  const beaconDeployment = await hre.deployments.deploy(name + "UpgradeBeacon", {
-    from: deployer.address,
-    args: [implementationAddress, controllerAddress],
-    log: true,
-    contract: "UpgradeBeacon",
-  });
-  const beaconAddress = beaconDeployment.address;
-  console.log(`${name} beaconAddress:`, beaconAddress);
+  let proxy: Contract;
+  let implementation: string;
+  let beaconAddress: string;
 
-  // Deploy proxy
-  const proxyDeployment = await hre.deployments.deploy(name + "UpgradeBeaconProxy", {
-    from: deployer.address,
-    args: [beaconAddress, initData],
-    log: true,
-    contract: "UpgradeBeaconProxy",
-  });
-  const proxyAddress = proxyDeployment.address;
-  console.log(`${name} proxyAddress:`, proxyAddress);
-  return hre.ethers.getContractAt(proxyDeployment.abi, proxyDeployment.address);
+  if (proxyDeployment) {
+    console.log(`${name} proxy deployed. upgrading........`);
+    beaconAddress = await hre.upgrades.erc1967.getBeaconAddress(proxyDeployment.address);
+    await hre.upgrades.upgradeBeacon(beaconAddress, factory);
+
+    console.log(`${name} proxy upgraded.`);
+
+    implementation = await hre.upgrades.beacon.getImplementationAddress(beaconAddress);
+    proxy = await hre.ethers.getContractAt(proxyDeployment.abi, proxyDeployment.address);
+  } else {
+    // Deploy new implementation
+    const beacon = await hre.upgrades.deployBeacon(factory);
```

Comment:
Really cool you can do this with the plugin!

Reply:
Sure.